### DISCUSSION TITLE
Accessing a computed Var in a Base object

### DISCUSSION THREAD

**Discussion by hpharmsen** [reflex-dev/reflex#3205] [Q&A] [answered]
I have a state which contains a list of objects.
How do I access a computed var in one of these sub objects?

To illustrate I simplified my code to the bare minimum:

`class Person(rx.Base):`
`    name: str`

`    @rx.var`
`    def hello_name(self) -> str:`
`        return "Hello " + self.name`

`class State(rx.State): `
`    count: int = 5`
`    persons: list[Person] = [Person(name='Alice'), Person(name='Bob')]`

`def index():`
`    return rx.foreach(`
`        State.persons,`
`        lambda person: rx.html('<p>' + person.hello_name + '</p>')`
`    )`

When I substitute hello_name for just name it works fine.
Somehow I don't get the concept.

**Comment by picklelo** [comment, marked as answer]
@hpharmsen The `@rx.var` decorator only works within state classes (not within the a Base class as you have it here). You want to use the `@property` decorator that's builtin to Python instead.

It looks like Pydantic supports [computed_fields](https://docs.pydantic.dev/2.0/usage/computed_fields/) from their v2.0 but We are still using 1.10 in Reflex.

Here's a way to do this currently by overriding the `dict` method to make sure the property gets serialized:

```python
import reflex as rx

class Person(rx.Base):
    name: str

    @property
    def hello_name(self) -> str:
        return "Hello " + self.name

    def dict(self, **kwargs):
        return super().dict(**kwargs) | {"hello_name": self.hello_name}


class State(rx.State):
    count: int = 5
    persons: list[Person] = [Person(name="Alice"), Person(name="Bob")]


@rx.page()
def index():
    return rx.foreach(
        State.persons, lambda person: rx.html("<p>" + person.hello_name + "</p>")
    )

app = rx.App()
```

**Reply by hpharmsen** [reply]
Thank you Nikhil. That works!

I presume this is not how things are normally supposed to work in Reflex. 
What would be the Reflex way of performing an operation on each element in a list which is part of the State?

**Reply by picklelo** [reply]
Normally the way you had it would work since each field in the class will get sent to the frontend. Pydantic doesn't by default send the `@property` tags down which is why we need this workaround. We can probably try to make this happen by default so you don't need to modify the dict method.

**Reply by hpharmsen** [reply]
Clear. Thank you!

**Comment by benedikt-bartscher** [comment]
Related to #2429, reflex could automatically serialize those properties without the `dict` workaround

**Comment by larsblumberg** [comment]
> Here's a way to do this currently by overriding the `dict` method to make sure the property gets serialized:
> 
> ```python
> import reflex as rx
> 
> class Person(rx.Base):
>     name: str
> 
>     @property
>     def hello_name(self) -> str:
>         return "Hello " + self.name
> 
>     def dict(self, **kwargs):
>         return super().dict(**kwargs) | {"hello_name": self.hello_name}
> 
> 
> class State(rx.State):
>     count: int = 5
>     persons: list[Person] = [Person(name="Alice"), Person(name="Bob")]
> 
> 
> @rx.page()
> def index():
>     return rx.foreach(
>         State.persons, lambda person: rx.html("<p>" + person.hello_name + "</p>")
>     )
> 
> app = rx.App()
> ```

I was using `def dict(self, **kwargs)` function on `rx.Model` classes. But now after upgrading from v0.8.13 -> v0.8.16 this `dict()` function seems no longer being used when serializing models to JSON objects?

Also with `rx.Model` being deprecated now, how can we expose custom computed fields to the frontend from `SQLModel` (or legacy `rx.Model`) classes?

**Reply by larsblumberg** [reply]
Ah, I found `serialize_sqlmodel()` in `model.py` which calls `.model_dump()`. I am overwriting this now instead of `.dict()` on `SQLModel` and it works as expected.

Would be great if there would be a (1) uniform (for all supported base classes) and (2) documented approach on this.

**Reply by masenf** [reply]
i believe it is both uniform and documented https://reflex.dev/docs/wrapping-react/serializers

What else do you think is missing?

**Reply by larsblumberg** [reply]
Hi Masen, thanks for pointing me to this documentation! Currently this document is organized in _"User Interface" > "Wrapping React" > "Serializers"_. Given that the serialization problem isn't stricly a "User Interface" nor a "React" problem how do  you feel about moving it to _"State" > "Vars" > "Serialization"_ to make it more accessible?
